Implementing highway net function

fukatani opened the pull request:
This PR will add `highway` network (https://arxiv.org/pdf/1505.00387v2.pdf).
This network is intended to shorten learning epoch of very deep network (>20 layer), and it is used for character language model. (https://arxiv.org/pdf/1508.06615v4.pdf)

Now, Forward / Backward calculation is not validated.

fukatani commented:
Hi!
I have implemented highway network.
Keras supported this network, so I think this feature is valuable to merge.

I'd appreciate it if I could get any opinion.

mattya commented:
Hi, thank you for PR!
Could you also update https://github.com/pfnet/chainer/blob/master/docs/source/reference/links.rst ?

mattya commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,79 @@
+        """
+        out_plain = self.activate(self.plain(x))
+        out_transform = sigmoid.sigmoid(self.transform(x))
+        out_transform_copy = copy.copy(out_transform)
```
Could you explain why you copy out_transform?

mattya commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,79 @@
+        out_transform = sigmoid.sigmoid(self.transform(x))
+        out_transform_copy = copy.copy(out_transform)
+        a = basic_math.mul(out_plain, out_transform)
+        ones = cuda.get_array_module(x).ones_like(x.data)
```
This code doesn't work on GPU.
  ones = cuda.get_array_module(x.data).ones_like(x.data) 
is correct.

mattya commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,79 @@
+        ones = cuda.get_array_module(x).ones_like(x.data)
+        b = basic_math.rsub(out_transform_copy, ones)
+        c = basic_math.mul(x, b)
+        y = basic_math.add(a, c)
```
I think 
`y = out_plain * out_transform + x * (1 - out_transform)`
is easy to understand.

fukatani commented:
Thanks for your attentive comments!

The reason why I copied 'out_transform' is the note of https://github.com/pfnet/chainer/blob/master/chainer/function.py as follows.
```
       As of v1.5, a function instance cannot be used twice in any
       computational graphs. In order to reuse a function object multiple
       times, use :func:`copy.copy` before the function applications to make a
       copy of the instance.
```
But I may misunderstand the meaning of it.

I added test for checking back-prop value and previous PR doesn't pass test.
So I fixed it to as you suggested (don't use copy).
And I reflected the other your pointing out (edit links.rst).

mattya commented:
Thank you for the update!  LGTM.

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,71 @@
+
+    """Highway module.
+
+    Network allowing unimpeded information flow across layer on information
```
Make it into a sentence, e.g. "This is a network fragment allowing ..."

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,71 @@
+
+    Network allowing unimpeded information flow across layer on information
+    flow.
+    It applies two different gate. One gate is transform gate, and the other
```
two different gate*s*

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,71 @@
+    Network allowing unimpeded information flow across layer on information
+    flow.
+    It applies two different gate. One gate is transform gate, and the other
+    is carry gate. Carry gate enable to propagete large gradient value to
```
Carry gate enable**s** **us** to ...

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,71 @@
+    previous layer.
+
+    The output is sum of linear multiplied transform gate, and input array
+    multiplied carry gate.
```
The output is a sum of the output of a fully-connected layer with the specified activation function multiplied by the transform gate, and the input array multiplied by the carry gate.

(Maybe this corrected sentence is too long, so it would be better to split it into multiple sentences.)

beam2d commented:
I added some comments on the document.
I also think that it is better to explicitly write down the equations that defines the forward computation of this chain.
If you need, I can take over the documentation task.

fukatani commented:
Thanke for your review!

I appreciate if you take over the documentation task.
My English is not natural,
repeating exchange about doc might be a time loss for you.

And I relaxed test `atol` and `rtol` for backword test.
Highway including product of two linear, 
I decided them to be square root of it in linear.

fukatani commented:
Because CI failed, I relaxed test `atol` and `rtol` for backword test.
Highway including product of two linear, 
I decided them to be square root of it in linear. (Or is there any reasonable method?)

mattya commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,79 @@
+    It applies two different gates. One gate is transform gate
+    (:math:`tanh(W_h x + b_h)`), and the other is carry gate
+    (:math:`sigma(W_t x + b_t)`). Carry gate enable us to propagete large
+    gradient value to previous layer.
```
One gate is ordinal non-linear transformation (:math:`activate(W_h x + b_h)`),
and the other is carry gate (:math:`1 - sigma(W_t x + b_t)`)

mattya commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,79 @@
+
+    .. math::
+
+    y &=& Relu(W_h x + b_h)\\odot\\sigma(W_t x + b_t)) +
```
Relu -> activate

fukatani commented:
Thank you for your review!
And I'm sorry for late response.
I fixed doc.

fukatani commented:
Thanks for mattya-san, document is fixed. Please check this PR.

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+        x \\odot(1 - \\sigma(W_t x + b_t))
+
+    The output array has the same spatial size as the input. In order to
+    satisfy this, :math:`W_h` and :math:`W_t` must be square matrix.
```
matrix -> matrices

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+        in_out_size (int): Dimension of input and output vectors.
+        nobias (bool): If ``True``, then this function does not use the bias.
+        activate: Activation function of plain array. :math:`tanh` is also
+        enable.
```
enable -> available (if I understand your intention correctly?)

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+        activate: Activation function of plain array. :math:`tanh` is also
+        enable.
+        init_Wh (2-D array): Initial weight value of plain array. If ``None``,
+            then this function uses to initialize ``wscale``.
```
uses to -> uses it to

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+            May also be a callable that takes ``numpy.ndarray`` or
+            ``cupy.ndarray`` and edits its value.
+        init_bh (1-D array): Initial bias value of plain array. If ``None``,
+            then this function uses to initialize zero vector.
```
ditto

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+            May also be a callable that takes ``numpy.ndarray`` or
+            ``cupy.ndarray`` and edits its value.
+        init_Wt (2-D array): Initial weight value of transform array.
+            If ``None``, then this function uses to initialize ``wscale``.
```
ditto

beam2d commented on chainer/links/connection/highway.py:
```diff
@@ -0,0 +1,75 @@
+            Default value is -1 vector.
+            May also be a callable that takes ``numpy.ndarray`` or
+            ``cupy.ndarray`` and edits its value.
+            Negative value is sufficient for learning (e.g. -1, -3, ...).
```
sufficient for learning -> recommended by the author of the paper? (if I understand your intention correctly)

beam2d commented:
I added some comments on the documentation. LGTM except those!

fukatani commented:
Thanks for your attentive review! :smiley:
I applied your comments.

mattya commented:
Thank you! 👍

beam2d commented:
LGTM!
